redwood-protocol-widget/src/commonMain/kotlin/app/cash/redwood/protocol/widget/ProtocolNode.kt

```diff
@@ -28,19 +28,27 @@ import app.cash.redwood.widget.Widget
  * @suppress
  */
 @RedwoodCodegenApi
-public interface ProtocolNode<W : Any> {
-  public val widget: Widget<W>
+public abstract class ProtocolNode<W : Any> {
+  public abstract val widget: Widget<W>
+
+  private var container: Widget.Children<W>? = null
```

Comment:
As the commit message details, soon this type will host `private val index: Int` so as to be able to remove and re-attach its `widget.view` within `container` at the correct position if the decorator modifiers change the instance.

Reply:
Who’s responsible for shifting 'em when an element is removed in the middle?

Reply:
For native Redwood we have a ChildrenNode which holds all of its WidgetNode children. For protocol Redwood we don't currently have that concept, and will need one.

Currently each ProtocolNode returns a Widget.Children when asked for a children container and the ProtocolBridge interacts with that directly. Next I'll instead return some wrapper around a Widget.Children that maintains the list of its ProtocolNode children and the Widgets into the Widget.Children. This way, when we do a tree operation, well ask this new type and it can update all the indices as well as perform the operation.